### PR TITLE
feat(log-store): add the object store WAL object format

### PR DESCRIPTION
**Pull request by fengjiachun**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

- RFC: #9195 (section *Object format*)
- Tracking issue: #9197

## What's changed and what's your intention?

This PR adds the byte format of a single object store WAL object to `log-store`. It is the object format slice of the tracking issue and does not depend on the other slices.

An object is laid out as:

```text
header | segment (region 1) | ... | segment (region N) | footer | trailer
```

- **Header.** It holds the magic `GTWALOBJ`, format version 1, the object sequence and a 16-byte writer instance id.
- **Segments.** Each segment holds the entries of one region, ordered by entry id. Segments are ordered by region id.
- **Footer.** For every segment it records the region id, the entry id range, the entry count, the byte range and a CRC32.
- **Trailer.** This fixed-size block holds the footer offset and length, the footer CRC32, the whole-object CRC32 and the magic `GTWALTRL`.

The encoder groups records by region and sorts them, so the same records always produce the same bytes. The RFC's identical-retry rule relies on that.

The header, trailer, footer and each segment decode separately, so a reader can find the footer from the trailer and decode one region's segment on its own. Decoding checks the magics, the format version, the footer and segment CRCs, and entry order. It also checks that the footer sits between the header and the trailer, and that the segment byte ranges cover the object body with no gaps or overlaps. A test-only reference decoder checks a whole object, including the whole-object CRC. Every failure returns the new `CorruptedWalObject` error, which maps to `StatusCode::Unexpected`.

The module has no callers yet. It is declared private with `#[allow(dead_code)]`. The batch, catalog, I/O and store modules that use it come in later PRs.

The PR adds `crc32fast` as a workspace dependency. The crate was already in the lockfile, so no new package versions come in.

This change does not break API or data compatibility, because it adds a new format and nothing reads or writes it yet.

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.
- [ ] This PR needs to be backported to release branches, and I have added the `backport-<target>` labels (e.g. `backport-v1.3` targets `release/v1.3`).

### CODE REVIEW

**Comment by fengjiachun** [comment]
@codex review

When you finish, append exactly this marker to the review summary:

**Comment by fengjiachun** [comment]
@codex review

When you finish, append exactly this marker to the review summary:

File: src/log-store/src/object_store_wal/format.rs [hunk 1]
```diff
@@ -0,0 +1,927 @@
...
+
+use bytes::{BufMut, Bytes, BytesMut};
+use snafu::{OptionExt, ensure};
+use store_api::storage::RegionId;
+
+use crate::error::{CorruptedWalObjectSnafu, Result};
+
+const HEADER_MAGIC: &[u8; 8] = b"GTWALOBJ";
+const TRAILER_MAGIC: &[u8; 8] = b"GTWALTRL";
+const FORMAT_VERSION: u16 = 1;
```
[the remark this answers is not in the document]
Comment (fengjiachun) [reply src/log-store/src/object_store_wal/format.rs:29]: Added `test_format_matches_version_1_fixture` in 9e615e99045cfaa8ae36ff5d340a7b10922ed0c0, which pins a 259-byte version 1 object derived from the documented layout and checks its exact decoding and byte-for-byte re-encoding, so a change to field order, endianness or checksum coverage now fails.

**Comment by fengjiachun** [comment]
@codex  review

File: src/log-store/src/object_store_wal/format.rs [hunk 2]
```diff
@@ -0,0 +1,1063 @@
...
+        expected_len == bytes.len(),
+        CorruptedWalObjectSnafu {
+            reason: format!(
+                "footer declares {count} entries, expected length {expected_len}, actual {}",
+                bytes.len()
+            ),
+        }
+    );
+
+    let mut directory = Vec::with_capacity(count);
```
Comment (WenyXu) [inline src/log-store/src/object_store_wal/format.rs:299]: `encode_object` rejects an empty record list, and the RFC defines every object as one sealed batch. However, `decode_footer` accepts `count == 0`, after which `verify_segment_ranges` also succeeds; only the test-only `decode_object` rejects an empty footer.

The future recovery path uses the production header/trailer/footer decoders, so it could accept a checksum-valid empty object that the writer never produces. Could we reject `count == 0` in `decode_footer` (or a shared production validation path), and add a crafted-footer test?

**Comment by WenyXu** [comment]
@codex review

File: src/log-store/src/object_store_wal/format.rs (hunk 2 above)
Comment (fengjiachun) [reply src/log-store/src/object_store_wal/format.rs:299]: Good catch, thanks. Fixed in 9ff5de9094e17810dfbc1de75e7f981a81bc2d63: `decode_footer` now rejects a zero entry count, and `test_format_rejects_checksum_valid_empty_footer` builds a checksum-valid object with an empty footer and checks that it is rejected.

File: src/log-store/src/object_store_wal/format.rs [hunk 3]
```diff
@@ -0,0 +1,1096 @@
...
+const SEGMENT_HEADER_LEN: usize = 8 + 4;
+/// Length of one footer entry: region id, entry id range, entry count, segment
+/// offset, segment length and segment CRC32.
+pub(super) const FOOTER_ENTRY_LEN: usize = 8 + 8 + 8 + 4 + 8 + 8 + 4;
+/// Length of the entry count the footer starts with.
+const FOOTER_COUNT_LEN: usize = 4;
+
+/// Header of a WAL object.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) struct Header {
```
Comment (WenyXu) [inline src/log-store/src/object_store_wal/format.rs]: This module introduces `pub(super)` for its entire parent-facing format API. We use `pub(crate)` for intra-crate module boundaries in `log-store` (for example, the Kafka implementation), rather than `pub(super)`.

Please use the established visibility convention here as well, or move the consuming code into the same module if the API should remain private.

**Review by WenyXu — APPROVED** [review]

File: src/log-store/src/object_store_wal/format.rs (hunk 3 above)
Comment (fengjiachun) [reply src/log-store/src/object_store_wal/format.rs]: Thanks, switched the format API from `pub(super)` to `pub(crate)` in bd41fe8c8486d52286fe6f2ff1e2a1338b6b0b1c.
